### PR TITLE
NAS-100718 / 11.3 / Misc improvements to mDNS plugin

### PR DESCRIPTION
**Pull request by anodos325**
- Move service mDNS advertisement to middleware
- Properly configure mixed AFP and SMB Time Machine shares. This will allow protocol failover from SMB to AFP.
- Bring in afp, ftp, iscsi, smb advertisement to middleware
- Persistently store VUID for Time Machine over AFP.
- Add ability to selectively restart mDNS advertisement for services (less disruptive than restarting everything).
- Add ability to which services we're currently advertising.
- Only advertise on Active Controller, and use virtual hostname on HA servers

### CODE REVIEW

File: src/middlewared/middlewared/plugins/afp.py [hunk 1]
```diff
@@ -109,6 +110,7 @@ class Config:
         UnixPerm('umask', default='000'),
         List('hostsallow', items=[], default=[]),
         List('hostsdeny', items=[], default=[]),
+        Str('vuid', null=True, default=''),
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/afp.py]: Do we need to expose vuid?

File: src/middlewared/middlewared/plugins/mdns.py [hunk 2]
```diff
@@ -1,14 +1,62 @@
...
+    FTP = 'mDNSServiceFTPThread'
+    HTTP = 'mDNSServiceHTTPThread'
+    HTTPS = 'mDNSServiceHTTPSThread'
+    ISCSI = 'mDNSServiceISCSIThread'
+    MIDDLEWARE = 'mDNSServiceMiddlewareThread'
+    MIDDLEWARE_SSL = 'mDNSServiceMiddlewareSSLThread'
+    NFS = 'mDNSServiceNFSThread'
+    SSH = 'mDNSServiceSSHThread'
+    SFTP_SSH = 'mDNSServiceSFTP_SSHThread'
+    SMB = 'mDNSServiceSMBThread'
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/mdns.py]: Is it a requirement to have a thread for each service?
If not, would it be too much work to have them in the same?

File: src/middlewared/middlewared/plugins/mdns.py [hunk 3]
```diff
@@ -231,6 +392,24 @@ def __init__(self, *args, **kwargs):
...
+    @private
+    async def get_hostname(self):
+        """
+        Return virtual hostname if the server is HA and this is the active controller.
+        Return None if this is the passive controller.
+        In all other cases return the hostname_local.
+        This is to ensure that the correct hostname is used for mdns advertisements.
+        """
+        ngc = await self.middleware.call('network.configuration.config')
+        if not await self.middleware.call('system.is_freenas') and await self.middleware.call('notfier.failover_licensed'):
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/mdns.py]: If think this check can be removed if we check for the existence of `hostname_virtual` in ngc?

File: src/middlewared/middlewared/plugins/mdns.py [hunk 4]
```diff
@@ -231,6 +392,24 @@ def __init__(self, *args, **kwargs):
...
+            failover_status = await self.middleware.call('notifier.failover_status')
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/mdns.py]: In 11.3 we use failover.status

File: src/middlewared/middlewared/plugins/mdns.py (hunk 3 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/mdns.py]: Yes. Good point

File: src/middlewared/middlewared/plugins/afp.py (hunk 1 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/afp.py]: Probably not, but we do it for SMB. It's nice to have the consistency. An advanced user may wish to convert a time machine share from AFP to SMB. In this case being able to transfer the VUID is useful.

File: src/middlewared/middlewared/plugins/mdns.py [hunk 5]
```diff
@@ -89,49 +120,201 @@ def start_mdnsd(self):
...
+        self.txtrecord = kwargs.get('txtrecord')
+        self.port = kwargs.get('port')
+        self.regtype = kwargs.get('regtype')
+        self.name = kwargs.get('name')
 
-    def active(self, sdRef):
-        return (bool(sdRef) and sdRef.fileno() != -1)
+    @property
+    def fullname(self):
+        return "%s.%s.%s" % (
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/mdns.py]: There are three %s and only two args.
Also lets use .format() or f-string please

File: src/middlewared/middlewared/plugins/mdns.py [hunk 6]
```diff
@@ -89,49 +120,201 @@ def start_mdnsd(self):
...
+            mDNSServices[srv.name].sdRef = pybonjour.DNSServiceRegister(
+                name=self.hostname,
+                regtype=self.regtype,
+                port=port,
+                txtRecord=txtrecord,
+                callBack=None
+            )
+
+        self.finished.wait()
+        self.logger.trace("we're finished")
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/mdns.py]: Is this a debug left-over or intended?

File: src/middlewared/middlewared/plugins/mdns.py [hunk 7]
```diff
@@ -240,36 +363,36 @@ def start(self):
...
-        for thread in self.threads.copy():
-            thread = self.threads.get(thread)
-            thread.cancel()
-            del self.threads[thread.service]
+        thread = self.threads.get('mDNSThread')
+        if thread is None:
+            return
+
+        thread.cancel()
+        del self.threads['mDNSThread']
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/mdns.py]: That seems unnecessary if we are setting threads below. The object will be garbage collected anyway.

File: src/middlewared/middlewared/plugins/service.py [hunk 8]
```diff
@@ -836,14 +837,17 @@ def _initializing_smartd_pid(self):
         # restarting netatalk.
         await self._system("pkill -9 afpd")
         await self._system("pkill -9 cnid_metad")
+        await self.middleware.call('mdnsadvertise.restart')
 
     async def _restart_afp(self, **kwargs):
         await self._stop_afp()
         await self._start_afp()
+        await self.middleware.call('mdnsadvertise.restart')
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/service.py]: this is already called on _start_afp

File: src/middlewared/middlewared/plugins/mdns.py (hunk 5 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/mdns.py]: I'll do better and replace where I was using this with a simple dictionary.

File: src/middlewared/middlewared/plugins/mdns.py (hunk 7 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/mdns.py]: Removed.

File: src/middlewared/middlewared/plugins/service.py (hunk 8 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/service.py]: Good observation. Removed.

**Pull request merged**